File: packages/client/native/src/pty.zig
// Flat, FFI-trivial pseudo-terminal over libc forkpty(3). Exposes exactly what
// the embedded Agent-tab terminal needs: spawn a child on a PTY, non-blocking
// read of its output, write, resize, and reap. Handles are small ints into a
// fixed table so bun:ffi only ever passes ints and byte pointers, never structs.
//
// Packed-argument contract (avoids shell parsing and per-token FFI marshalling):
//   argv_packed: "arg0\0arg1\0...\0"  - NUL-separated, terminated by an empty
//                                       token (a trailing extra NUL).
//   env_packed:  "K=V\0K=V\0...\0"    - same shape; empty means inherit parent.
//   cwd:         a plain C string, or empty to keep the parent's directory.

const std = @import("std");
const builtin = @import("builtin");

const c = @cImport({
    @cInclude("unistd.h");
    @cInclude("fcntl.h");
    @cInclude("errno.h");
    @cInclude("signal.h");
    @cInclude("stdlib.h");
    @cInclude("string.h");
    @cInclude("termios.h");
    @cInclude("sys/ioctl.h");
    @cInclude("sys/wait.h");
    if (builtin.os.tag == .macos) @cInclude("util.h") else @cInclude("pty.h");
});

extern var environ: [*c][*c]u8;

const MAX_SESSIONS = 64;
const MAX_TOKENS = 1024;

const PtySession = struct {
    in_use: bool = false,
    master_fd: c_int = -1,
    pid: c.pid_t = 0,
    reaped: bool = false,
    exit_code: c_int = 0,
};

var sessions = [_]PtySession{.{}} ** MAX_SESSIONS;

// Split a double-NUL-terminated packed buffer into a NULL-terminated array of
// pointers into that same buffer. Returns the token count, or -1 on overflow.
fn splitPacked(packed_buffer: [*c]u8, out: [][*c]u8) c_int {
    var count: usize = 0;
    var cursor: [*c]u8 = packed_buffer;
    while (cursor[0] != 0) {
        if (count >= out.len - 1) return -1;
        out[count] = cursor;
        count += 1;
        cursor += c.strlen(cursor) + 1;
    }
    out[count] = null;
    return @intCast(count);
}

// Reap the child once, recording its exit status. WNOHANG unless `block`.
fn reapSession(session: *PtySession, block: bool) void {
    if (session.reaped) return;
    var status: c_int = 0;
    const result = c.waitpid(session.pid, &status, if (block) 0 else c.WNOHANG);
    if (result != session.pid) return;
    session.reaped = true;
    if (c.WIFEXITED(status)) {
        session.exit_code = c.WEXITSTATUS(status);
    } else if (c.WIFSIGNALED(status)) {
        session.exit_code = 128 + c.WTERMSIG(status);
    } else {
        session.exit_code = -1;
    }
}

fn sessionOf(handle: c_int) ?*PtySession {
    if (handle < 0 or handle >= MAX_SESSIONS) return null;
    const session = &sessions[@intCast(handle)];
    if (!session.in_use) return null;
    return session;
}

// Spawn `argv_packed` on a fresh PTY sized cols x rows, in cwd, with env_packed
// (or the parent's env when empty). Returns a session handle, or -1 on failure.
export fn cueloop_pty_spawn(argv_packed: [*c]u8, cwd: [*c]const u8, env_packed: [*c]u8, cols: c_int, rows: c_int) c_int {
    var handle: c_int = -1;
    for (0..MAX_SESSIONS) |index| {
        if (!sessions[index].in_use) {
            handle = @intCast(index);
            break;
        }
    }
    if (handle < 0) return -1;

    var argv: [MAX_TOKENS][*c]u8 = undefined;
    if (splitPacked(argv_packed, &argv) < 1) return -1;

    var envp: [MAX_TOKENS][*c]u8 = undefined;
    const have_env = env_packed != null and env_packed[0] != 0;
    if (have_env and splitPacked(env_packed, &envp) < 0) return -1;

    var winsize = std.mem.zeroes(c.struct_winsize);
    winsize.ws_col = @intCast(cols);
    winsize.ws_row = @intCast(rows);

    var master_fd: c_int = -1;
    const pid = c.forkpty(&master_fd, null, null, &winsize);
    if (pid < 0) return -1;

    if (pid == 0) {
        // Child: adopt cwd/env then exec; _exit on any failure so a broken child never runs JS.
        if (cwd != null and cwd[0] != 0) {
            if (c.chdir(cwd) != 0) c._exit(127);
        }
        if (have_env) environ = &envp;
        _ = c.execvp(argv[0], &argv);
        c._exit(127);
    }

    // Parent: non-blocking master so reads can poll without stalling the loop.
    const flags = c.fcntl(master_fd, c.F_GETFL, @as(c_int, 0));
    _ = c.fcntl(master_fd, c.F_SETFL, flags | c.O_NONBLOCK);

    sessions[@intCast(handle)] = .{
        .in_use = true,
        .master_fd = master_fd,
        .pid = pid,
        .reaped = false,
        .exit_code = 0,
    };
    return handle;
}

export fn cueloop_pty_write(handle: c_int, data: [*c]const u8, len: c_int) c_int {
    const session = sessionOf(handle) orelse return -1;
    const written = c.write(session.master_fd, data, @intCast(len));
    return @intCast(written);
}

// Non-blocking read of child output. Returns bytes read (>0), 0 when nothing is
// available yet, -2 once the child has exited and its output is drained, or -1.
export fn cueloop_pty_read(handle: c_int, buf: [*c]u8, len: c_int) c_int {
    const session = sessionOf(handle) orelse return -1;
    const bytes_read = c.read(session.master_fd, buf, @intCast(len));
    if (bytes_read > 0) return @intCast(bytes_read);
    if (bytes_read == 0) {
        reapSession(session, true);
        return -2; // EOF: slave closed
    }
    const err = std.c._errno().*;
    if (err == c.EAGAIN or err == c.EWOULDBLOCK) {
        reapSession(session, false);
        return if (session.reaped) -2 else 0; // exited with no more output vs still running
    }
    if (err == c.EIO) {
        reapSession(session, true);
        return -2; // Linux: slave gone
    }
    return -1;
}

export fn cueloop_pty_resize(handle: c_int, cols: c_int, rows: c_int) c_int {
    const session = sessionOf(handle) orelse return -1;
    var winsize = std.mem.zeroes(c.struct_winsize);
    winsize.ws_col = @intCast(cols);
    winsize.ws_row = @intCast(rows);
    return c.ioctl(session.master_fd, c.TIOCSWINSZ, &winsize);
}

export fn cueloop_pty_kill(handle: c_int) c_int {
    const session = sessionOf(handle) orelse return -1;
    return c.kill(session.pid, c.SIGTERM);
}

export fn cueloop_pty_get_pid(handle: c_int) c_int {
    const session = sessionOf(handle) orelse return -1;
    return @intCast(session.pid);
}

export fn cueloop_pty_get_exit_code(handle: c_int) c_int {
    const session = sessionOf(handle) orelse return -1;
    reapSession(session, true);
    return session.exit_code;
}

export fn cueloop_pty_close(handle: c_int) void {
    const session = sessionOf(handle) orelse return;
    // Close the master first (hangs up the child), then block to reap it: after the
    // SIGTERM from kill() plus this SIGHUP a well-behaved child exits at once, so the
    // wait returns promptly and never leaves a zombie.
    if (session.master_fd >= 0) _ = c.close(session.master_fd);
    reapSession(session, true);
    session.in_use = false;
    session.master_fd = -1;
}

File: packages/client/native/src/pty.c
// A flat, FFI-trivial pseudo-terminal over libc forkpty(3) - the code cueloop
// owns in place of the third-party bun-pty package. It exposes exactly what the
// embedded Agent-tab terminal needs: spawn a child on a PTY, non-blocking read
// of its output, write, resize, and reap. Handles are small ints into a fixed
// table so bun:ffi only ever passes ints and byte pointers (never structs).
//
// Packed-argument contract (avoids shell parsing and per-token FFI marshalling):
//   argv_packed: "arg0\0arg1\0...\0"  - NUL-separated, terminated by an empty
//                                       token (a trailing extra NUL).
//   env_packed:  "K=V\0K=V\0...\0"    - same shape; empty means inherit parent.
//   cwd:         a plain C string, or empty to keep the parent's directory.

#include <stdint.h>
#include <stddef.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <signal.h>
#include <sys/ioctl.h>
#include <sys/wait.h>
#include <termios.h>
#if defined(__APPLE__)
#include <util.h>
#else
#include <pty.h>
#endif

extern char **environ;

#define MAX_SESSIONS 64
#define MAX_TOKENS 1024

typedef struct {
  int in_use;
  int master_fd;
  pid_t pid;
  int reaped;
  int exit_code;
} PtySession;

static PtySession sessions[MAX_SESSIONS];

// Split a double-NUL-terminated packed buffer into a NULL-terminated array of
// pointers into that same buffer. Returns the token count, or -1 on overflow.
static int split_packed(char *packed, char **out, int max) {
  int count = 0;
  char *cursor = packed;
  while (*cursor != '\0') {
    if (count >= max - 1) return -1;
    out[count++] = cursor;
    cursor += strlen(cursor) + 1;
  }
  out[count] = NULL;
  return count;
}

// Reap the child once, recording its exit status. WNOHANG unless `block`.
static void reap_session(PtySession *s, int block) {
  if (s->reaped) return;
  int status = 0;
  pid_t result = waitpid(s->pid, &status, block ? 0 : WNOHANG);
  if (result != s->pid) return;
  s->reaped = 1;
  if (WIFEXITED(status)) s->exit_code = WEXITSTATUS(status);
  else if (WIFSIGNALED(status)) s->exit_code = 128 + WTERMSIG(status);
  else s->exit_code = -1;
}

// Spawn `argv_packed` on a fresh PTY sized cols x rows, in cwd, with env_packed
// (or the parent's env when empty). Returns a session handle, or -1 on failure.
int cueloop_pty_spawn(char *argv_packed, const char *cwd, char *env_packed,
                      int cols, int rows) {
  int handle = -1;
  for (int i = 0; i < MAX_SESSIONS; i++) {
    if (!sessions[i].in_use) { handle = i; break; }
  }
  if (handle < 0) return -1;

  char *argv[MAX_TOKENS];
  if (split_packed(argv_packed, argv, MAX_TOKENS) < 1) return -1;

  char *envp[MAX_TOKENS];
  int have_env = env_packed != NULL && env_packed[0] != '\0';
  if (have_env && split_packed(env_packed, envp, MAX_TOKENS) < 0) return -1;

  struct winsize ws = {0};
  ws.ws_col = (unsigned short)cols;
  ws.ws_row = (unsigned short)rows;

  int master_fd = -1;
  pid_t pid = forkpty(&master_fd, NULL, NULL, &ws);
  if (pid < 0) return -1;

  if (pid == 0) {
    // Child: adopt the requested cwd/env, then become the target program. On any
    // failure exec never returns, so _exit keeps a broken child from running JS.
    if (cwd && cwd[0] != '\0') { if (chdir(cwd) != 0) _exit(127); }
    if (have_env) environ = envp;  // execvp resolves PATH from environ
    execvp(argv[0], argv);
    _exit(127);
  }

  // Parent: non-blocking master so reads can poll without stalling the loop.
  int flags = fcntl(master_fd, F_GETFL, 0);
  fcntl(master_fd, F_SETFL, flags | O_NONBLOCK);

  sessions[handle].in_use = 1;
  sessions[handle].master_fd = master_fd;
  sessions[handle].pid = pid;
  sessions[handle].reaped = 0;
  sessions[handle].exit_code = 0;
  return handle;
}

static PtySession *session_of(int handle) {
  if (handle < 0 || handle >= MAX_SESSIONS || !sessions[handle].in_use) return NULL;
  return &sessions[handle];
}

int cueloop_pty_write(int handle, const uint8_t *data, int len) {
  PtySession *s = session_of(handle);
  if (!s) return -1;
  ssize_t written = write(s->master_fd, data, (size_t)len);
  return (int)written;
}

// Non-blocking read of child output. Returns bytes read (>0), 0 when nothing is
// available yet, -2 once the child has exited and its output is drained, or -1.
int cueloop_pty_read(int handle, uint8_t *buf, int len) {
  PtySession *s = session_of(handle);
  if (!s) return -1;
  ssize_t n = read(s->master_fd, buf, (size_t)len);
  if (n > 0) return (int)n;
  if (n == 0) { reap_session(s, 1); return -2; }  // EOF: slave closed
  if (errno == EAGAIN || errno == EWOULDBLOCK) {
    reap_session(s, 0);
    return s->reaped ? -2 : 0;  // exited with no more output vs still running
  }
  if (errno == EIO) { reap_session(s, 1); return -2; }  // Linux: slave gone
  return -1;
}

int cueloop_pty_resize(int handle, int cols, int rows) {
  PtySession *s = session_of(handle);
  if (!s) return -1;
  struct winsize ws = {0};
  ws.ws_col = (unsigned short)cols;
  ws.ws_row = (unsigned short)rows;
  return ioctl(s->master_fd, TIOCSWINSZ, &ws);
}

int cueloop_pty_kill(int handle) {
  PtySession *s = session_of(handle);
  if (!s) return -1;
  return kill(s->pid, SIGTERM);
}

int cueloop_pty_get_pid(int handle) {
  PtySession *s = session_of(handle);
  return s ? (int)s->pid : -1;
}

int cueloop_pty_get_exit_code(int handle) {
  PtySession *s = session_of(handle);
  if (!s) return -1;
  reap_session(s, 1);
  return s->exit_code;
}

void cueloop_pty_close(int handle) {
  PtySession *s = session_of(handle);
  if (!s) return;
  if (s->master_fd >= 0) close(s->master_fd);
  reap_session(s, 0);
  s->in_use = 0;
  s->master_fd = -1;
}
